### PR TITLE
Base model

### PR DESCRIPTION
**Pull request by MeltingShoe**
I've created a "class factory" function that constructs model classes. It uses typeclass construction instead of the normal method, but it doesn't really change how it works it was just easier to do it that way. I've consolidated the file structure so there's a single file defining all ff classes, training functions, and perform action functions. Train and perform_action have been changed to functions rather than classes because it makes construction easier and there's no reason for them to be classes(as far as I can tell, correct me if I'm wrong). The params that train used to take are now taken in the base class instead. Feed_forward is still it's own class because pytorch pretty much requires it. Overall I'm pretty happy with this because I don't think we'll really have to touch the base model class again once we've made a runtime that works with the gym environment(fingers crossed).

Issues:

- We should define logic for setting the filepath dynamically in BaseModel rather than the runtime.
- I think cuda may be broken. My gpu usage still spikes during every epoch but my CPU/RAM usage spikes in sync with it. It might be that my CPU is handling either part of the training/validation computation or it might just be from saving the checkpoint file. 30% usage on a Ryzen 1700 seems fairly high just to save a file though. I probably deleted some code inside default_train or supervised_evaluate that enables cuda without realizing it had functionality.
- We'll have to change some of the internal logic to allow for perform_action once we've built a chess runtime
- The interfaces I made don't do anything because the names of the abstract methods are defined explicitly. We might want a check that the correct functions are fed into the generate_class() function.
- A lot of things could probably still be cleaned up.

### CODE REVIEW

**Comment by MeltingShoe** [comment]
[Base model](https://trello.com/c/ZDnyAPdp/9-base-model)

**Comment by MeltingShoe** [comment]
[cifar10 example](https://trello.com/c/e2PLtAYp/10-cifar10-example)

File: classes/base_model.py [hunk 1]
```diff
@@ -1,30 +1,65 @@
...
+    '''
+    Setting BaseModel doesn't actually do anything atm because we
+    explicitely define the abstract methods in here
+    I don't think it really matters though because explicitely defining the
+    methods should prevent any inheritance issues
+    '''
+    superclasses = (BaseModel,)
+
+    '''
+    I felt that defining training params inside the model was cleaner and
```
Comment (achaiah) [inline classes/base_model.py:31]: Just for clarification, which parameters are we talking about?  Is it the parameters of the model (e.g. number of layers, types of layers, etc) or is it metaparameters (e.g. learning rate, optimizer etc)?  If it's the latter then I suggest we do not keep parameters with the model as those are logically separate things.

**Comment by achaiah** [comment]
"Feed_forward is still it's own class because pytorch pretty much requires it."  -- Could you elaborate on that? The networks I've used before with pytorch don't require it so just curious as to why this is different.

File: classes/base_model.py (hunk 1 above)
Comment (MeltingShoe) [reply classes/base_model.py:31]: We talked about this a lot in the meeting today. Basically I think calling it a model class is a misnomer and should instead be thought of as a container for a single experiment. Metaparameters are logically separate from the model, but in the context of each generated class being a container for a single experiment I think it makes more sense to store metaparameters inside that container. Sorry for the naming confusion.

**Comment by MeltingShoe** [comment]
I'll quote cece95 here:

> You need a class because the architecture of the model and the FF function itself are strictly connected (every architecture needs its on FF function) and pytorch needs the architecture and the function to be defined in a class that extends pytorch module

Basically as far as I can tell in order for pytorch to recognize the FF computation as a nn.Module it needs to be it's own class with the trainable params set in it's init for it do be able to build the computation graph properly and for autograd to work. There could be some workaround, but because it's base has functionality that allows you to call the forward method by just calling the class I don't think it really affects anything. I may be wrong in saying that it requires it but I've found this is what works and don't really see any downsides to this approach.
